tests/test_dsl.py

```diff
@@ -320,3 +327,321 @@ def assert_cohorts_equivalent(dsl_cohort, qm_cohort):
     # ...and if the columns are the same.
     for k in dsl_col_defs:
         assert repr(dsl_col_defs[k]) == repr(qm_col_defs[k])
+
+
+def test_boolseries_and():
+    series1 = BoolSeries(ValueFromRow(source=None, column="test1"))
+    series2 = BoolSeries(ValueFromRow(source=None, column="test2"))
+
+    output = series1 & series2
+    assert isinstance(output, BoolSeries)
+
+    # do we get the expected Comparator structure?
+    assert output.value.connector == "and_"
+    assert not output.value.negated
+    assert output.value.operator is None
+
+    # ValueFromRow & ValueFromRow generates a Comparator for each side of the
+    # __and__ before the BoolSeries __and__ takes over, make sure our inner
+    # Comparators are constructed as expected.
+    assert isinstance(output.value.lhs, Comparator)
+    assert output.value.lhs.lhs.column == "test1"
+
+    assert isinstance(output.value.rhs, Comparator)
+    assert output.value.rhs.lhs.column == "test2"
```

Comment:
I think I'd prefer tests that look like:

```
def test_boolseries_and():
    series1 = BoolSeries(ValueFromRow(source=None, column="test1"))
    series2 = BoolSeries(ValueFromRow(source=None, column="test2"))

    exp_value = Comparator(
        connector='and_', 
        negated=False, 
        operator=None,
        lhs=Comparator(lhs=series1),
        rhs=Comparator(lhs=series2),
    )

    assert series1 & series2 ==BoolSeries(value=exp_value)
```

A few reasons:

1. Documentation: the behaviour is, by necessity, split up into lots of different functions, and it's helpful to have some tests that show exactly what gets created when eg you combine two series.  I think I'd find it easier to understand tests where the output is spelt out in full, rather than where each different part of the output is checked separately.
2. Maintainability: if we change the way that the output values are constructed, then it'll be fiddly to update each of these small assertions.
3. Coverage + correctness: from reading this, I don't know whether there are any attributes of `output` that are not tested.  There's no coverage-like tool that I know of for checking this, but I think it's pretty important for correctness.

Reply:
This captures my main feedback on these new tests perfectly.

Reply:
One more thought on testing: should the DSL tests know anything about the Query Model?

Reply:
@inglesp – updated relevant tests, new version is much nicer, thanks!

---

> should the DSL tests know anything about the Query Model?

Ben and I discussed this when I was starting the tests and came to the outcome that the DSL produces query model objects so the tests should reflect that.

Reply:
I think they should. In my mind the Query Model is the core of the system. The DSL exists to create the Query Model. Ultimately I think that all the DSL tests should be assertions of the query model graphs they produce.

Reply:
I wasn't clear, sorry.  I think we all agree that the tests should be of the form `assert process(dsl_obj) == qm_obj`.

However, `dsl_obj` is created with knowledge of the Query Model -- for instance `BoolSeries(ValueFromRow(source=None, column="test1"))`.

I'm not sure this is right, because we might end up with tests that create objects that are never actually createable through the DSL.

Reply:
I see, thank you.

So one option is that all the tests just use the facilities available to end users. But then we get much bigger QM graphs than we need for a given test and have to assert against the whole thing (making it over-specified and a bit fragile to change).

Or could do as above and come up with some generic traversal/pattern-matching tooling to pick out just the bit of the graph we actually need to assert against.

Or a smaller change would be to have something like this:

```py
def test_boolseries_and():
    series1 = BoolSeries(a_value())
    series2 = BoolSeries(a_value())

    assert series1 & series2 == BoolSeries(value=Comparator(
        connector='and_', 
        negated=False, 
        operator=None,
        lhs=Comparator(lhs=series1),
        rhs=Comparator(lhs=series2),
    ))
```

Where we hide the value-construction stuff to emphasise that it's not important and make it easier to change if necessary.

Reply:
Hiding the value construction is a good solution -- it gives readable tests while ensuring that we only use valid constructions.